In [83]:
import pandas as pd
import matplotlib.pyplot as plt

## Cleaning the raw files
****

### Anime information
important data about individual anime providing sufficient information about trends in time about important aspects of anime
****
Columns explained:

    animeID: id of anime as in anime url https://myanimelist.net/anime/1
    name: title of anime
    premiered: premiered on. default format (season year)
    genre: list of genre
    type: type of anime (example TV, Movie etc)
    episodes: number of episodes
    studios: list of studio
    source: source of anime (example original, manga, game etc)
    scored: score of anime
    scoredBy: number of member scored the anime
    members: number of member added anime to their list

Determined unused columns:
- title_english
- title_japanese
- title_synonyms
- image_url
- type: TV,OVA,...
- aired_string: just string version of aired column's dict(contains from and to dates)
- duration
- background
- broadcast : time the manga is broadcasted
- premiered
- opening_theme
- ending_theme
- premiered
- licensor
- _source_? 
- _members_?

In [84]:
anime_info_raw = pd.read_csv('myanimelist/AnimeList.csv')
anime_info_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14478 entries, 0 to 14477
Data columns (total 31 columns):
anime_id          14478 non-null int64
title             14478 non-null object
title_english     5724 non-null object
title_japanese    14443 non-null object
title_synonyms    8937 non-null object
image_url         14382 non-null object
type              14478 non-null object
source            14478 non-null object
episodes          14478 non-null int64
status            14478 non-null object
airing            14478 non-null bool
aired_string      14478 non-null object
aired             14478 non-null object
duration          14478 non-null object
rating            14478 non-null object
score             14478 non-null float64
scored_by         14478 non-null int64
rank              12904 non-null float64
popularity        14478 non-null int64
members           14478 non-null int64
favorites         14478 non-null int64
background        1057 non-null object
premiered         40

In [157]:
anime_info_raw.groupby('type').count()

anime_id  title  title_english  title_japanese  title_synonyms  \
type                                                                      
Movie        2624   2624           1176            2608            1681   
Music         849    849            162             845             603   
ONA          1144   1144            326            1140             610   
OVA          3612   3612           1122            3609            1858   
Special      1948   1948            496            1944            1434   
TV           4271   4271           2440            4267            2742   
Unknown        30     30              2              30               9   

         image_url  source  episodes  status  airing  ...  background  \
type                                                  ...               
Movie         2569    2624      2624    2624    2624  ...         267   
Music          849     849       849     849     849  ...          16   
ONA           1142    1144      1144    1144    1144  ...          74   
OVA           3585    3612      3612    3612    3612  ...         102   
Special       1942    1948      1948    1948    1948  ...          70   
TV            4265    4271      4271    4271    4271  ...         528   
Unknown         30      30        30      30      30  ...           0   

         premiered  broadcast  related  producer  licensor  studio  genre  \
type                                                                        
Movie            0          0     2624      1102       499    1293   2581   
Music            0          0      849       499         2     136    849   
ONA              0          0     1144       405        51     505   1134   
OVA              0          0     3612      2272       721    1753   3607   
Special          0          0     1948       838       383    1339   1947   
TV            4096       4271     4271      3168      1717    3507   4267   
Unknown          0          0       30         4         0      11     29   

         opening_theme  ending_theme  
type                                  
Movie             2624          2624  
Music              849           849  
ONA               1144          1144  
OVA               3612          3612  
Special           1948          1948  
TV                4271          4271  
Unknown             30            30  

[7 rows x 30 columns]

In [133]:
anime_info_raw.groupby('source').count()

anime_id  title  title_english  title_japanese  title_synonyms  \
source                                                                          
4-koma manga        224    224             95             224             138   
Book                 93     93             46              93              58   
Card game            59     59             22              58              26   
Digital manga        10     10              6              10               9   
Game                607    607            283             605             383   
Light novel         563    563            355             562             441   
Manga              3120   3120           1470            3117            2119   
Music               314    314             77             313             208   
Novel               359    359            198             359             268   
Original           3368   3368           1391            3351            1965   
Other               408    408            123             408             219   
Picture book         96     96             32              96              54   
Radio                 9      9              1               9               6   
Unknown            4210   4210           1245            4202            2519   
Visual novel        888    888            309             886             427   
Web manga           150    150             71             150              97   

               image_url  type  episodes  status  airing  ...  background  \
source                                                    ...               
4-koma manga         223   224       224     224     224  ...          25   
Book                  93    93        93      93      93  ...          17   
Card game             59    59        59      59      59  ...           1   
Digital manga         10    10        10      10      10  ...           1   
Game                 607   607       607     607     607  ...          52   
Light novel          563   563       563     563     563  ...         121   
Manga               3119  3120      3120    3120    3120  ...         305   
Music                313   314       314     314     314  ...          11   
Novel                359   359       359     359     359  ...          56   
Original            3323  3368      3368    3368    3368  ...         305   
Other                387   408       408     408     408  ...          36   
Picture book          93    96        96      96      96  ...           9   
Radio                  9     9         9       9       9  ...           1   
Unknown             4186  4210      4210    4210    4210  ...          65   
Visual novel         888   888       888     888     888  ...          39   
Web manga            150   150       150     150     150  ...          13   

               premiered  broadcast  related  producer  licensor  studio  \
source                                                                     
4-koma manga         118        123      224       158        72     198   
Book                  30         30       93        50        11      45   
Card game             37         38       59        35        10      48   
Digital manga          6          7       10         5         2       7   
Game                 195        210      607       334       204     440   
Light novel          277        285      563       402       354     527   
Manga               1241       1263     3120      2107      1178    2543   
Music                 11         14      314       218         4      56   
Novel                121        129      359       234       109     293   
Original             924        963     3368      1722       712    1839   
Other                 98        107      408       154        50     172   
Picture book          43         45       96        46         3      54   
Radio                  4          4        9         2         2       5   
Unknown              778     

In [134]:
anime_info_raw.groupby('episodes').count().sort_values(ascending=True,by='episodes').head() # 504 anime with 0 episodes!


anime_id  title  title_english  title_japanese  title_synonyms  \
episodes                                                                   
0              504    504            105             501             206   
1             6857   6857           2199            6833            4207   
2             1187   1187            373            1186             651   
3              558    558            191             557             363   
4              373    373            124             373             219   

          image_url  type  source  status  airing  ...  background  premiered  \
episodes                                           ...                          
0               501   504     504     504     504  ...          26        184   
1              6773  6857    6857    6857    6857  ...         389          0   
2              1184  1187    1187    1187    1187  ...          31          5   
3               558   558     558     558     558  ...          24         10   
4               370   373     373     373     373  ...          11         20   

          broadcast  related  producer  licensor  studio  genre  \
episodes                                                          
0               272      504       173        18     233    501   
1                 0     6857      3132       943    3149   6804   
2                 5     1187       780       232     616   1186   
3                10      558       348       138     294    556   
4                21      373       243        94     226    372   

          opening_theme  ending_theme  
episodes                               
0                   504           504  
1                  6857          6857  
2                  1187          1187  
3                   558           558  
4                   373           373  

[5 rows x 30 columns]

In [145]:
# exploring the 504 anime with 0 epsides 
anime_info_raw[(anime_info_raw['episodes'] == 0)][['score','status','title']]
# some still have score -> this means that the episodes value is not accurate: episodes have aired but episode count
# data has not been updated


score            status  \
31      7.73  Currently Airing   
34      8.54  Currently Airing   
67      3.44  Currently Airing   
145     0.00     Not yet aired   
187     0.00     Not yet aired   
198     7.50  Currently Airing   
334     0.00     Not yet aired   
387     0.00     Not yet aired   
402     5.26  Currently Airing   
458     0.00     Not yet aired   
463     7.20  Currently Airing   
502     4.57  Currently Airing   
511     0.00     Not yet aired   
669     5.54  Currently Airing   
671     0.00     Not yet aired   
687     6.13  Currently Airing   
724     4.41  Currently Airing   
733     4.29  Currently Airing   
750     7.31  Currently Airing   
833     4.58   Finished Airing   
838     5.46  Currently Airing   
864     0.00     Not yet aired   
877     4.80  Currently Airing   
890     6.89  Currently Airing   
947     0.00     Not yet aired   
955     4.50  Currently Airing   
988     0.00     Not yet aired   
1062    6.73  Currently Airing   
1072    7.60  Currently Airing   
1154    4.40  Currently Airing   
...      ...               ...   
14272   0.00     Not yet aired   
14275   0.00     Not yet aired   
14280   4.00  Currently Airing   
14282   5.00  Currently Airing   
14287   0.00  Currently Airing   
14289   0.00     Not yet aired   
14290   0.00     Not yet aired   
14291   0.00     Not yet aired   
14292   5.60  Currently Airing   
14296   0.00     Not yet aired   
14352   7.88  Currently Airing   
14368   0.00     Not yet aired   
14371   0.00     Not yet aired   
14374   0.00     Not yet aired   
14379   7.80  Currently Airing   
14383   5.25   Finished Airing   
14387   0.00  Currently Airing   
14391   7.40  Currently Airing   
14398   7.00  Currently Airing   
14403   9.52  Currently Airing   
14404   0.00     Not yet aired   
14408   0.00  Currently Airing   
14414   0.00     Not yet aired   
14416   0.00     Not yet aired   
14442   0.00  Currently Airing   
14449   5.00  Currently Airing   
14459   4.50  Currently Airing   
14460   3.00  Currently Airing   
14463   0.00     Not yet aired   
14464   4.57  Currently Airing   

                                                   title  
31                                      Crayon Shin-chan  
34                                             One Piece  
67                                         Neko Konogoro  
145                                      Carol & Tuesday  
187                                   Sidonia no Kishi 3  
198                            Ore ga Kanojo wo *su Wake  
334                        Shoujo☆Kageki Revue Starlight  
387                                        Mo Dao Zu Shi  
402         Amai Choubatsu: Watashi wa Kanshu Senyou Pet  
458                                             W&#039;z  
463                        Uma Musume: Pretty Derby (TV)  
502                                     Usagi no Matthew  
511       Fate/Extra: Last Encore - Irusterias Tendouron  
669                                         Neko no Robu  
671                                          Kori no Koe  
687                                        Saimin☆Gakuen  
724                                         Nineko Gelée  
733                                 Lord of Vermilion IV  
750                             Chibi Maruko-chan (1995)  
833                                           Happy Jozy  
838                                Beatless Intermission  
864                                Yakusoku no Neverland  
877                                         Konigiri-kun  
890                             Gegege no Kitarou (2018)  
947                        Sword Art Online: Alicization  
955                                     Unicorn no Kyupi  
988                                 Strike the Blood III  
1062                                    Granblue Fantasy  
1072                                      Major 2nd (TV)  
1154                                          SNS Police  
...                                                  ...  
14272     

In [148]:
anime_info_raw[(anime_info_raw['episodes'] == 0) & (anime_info_raw['status'] == 'Currently Airing') & (anime_info_raw['score'] == 0.00)][['score','status','title','rank','scored_by']]

score            status                                title     rank  \
6688     0.0  Currently Airing        Neko Neko Nihonshi 3rd Season  11431.0   
11765    0.0  Currently Airing                         Working Boys  12679.0   
14287    0.0  Currently Airing  Chibi Airin no Yuruyaka na Nichijou   9878.0   
14387    0.0  Currently Airing                   Washimo 6th Season  12690.0   
14408    0.0  Currently Airing     W・TAILS CAT - A STRANGE PRESENCE      0.0   
14442    0.0  Currently Airing          Chara to Otamajakushi Shima   9865.0   

       scored_by  
6688           0  
11765          0  
14287          0  
14387          0  
14408          0  
14442          0

In [154]:
anime_info_raw[anime_info_raw['scored_by'] == 0][['anime_id','title','status','score','scored_by']] 
# exists anime that have not been scored by any user!

anime_id                                              title  \
55        33185           Peace Maker Kurogane Movie 1: Omou Michi   
76        37447                                    Karakuri Circus   
109       36951                            Yuu☆Yuu☆Hakusho Special   
145       37435                                    Carol & Tuesday   
187       35759                                 Sidonia no Kishi 3   
229       34779                     Ar:pieL - 6-seosgaeui Unmyeong   
334       35503                      Shoujo☆Kageki Revue Starlight   
387       37208                                      Mo Dao Zu Shi   
458       37509                                           W&#039;z   
639       33248             K: Seven Stories Movie 1 - R:B - Blaze   
671       37106                                        Kori no Koe   
864       37779                              Yakusoku no Neverland   
894       35424  Donten ni Warau Gaiden: Shukumei, Soutou no Fuuma   
930       37303  K: Seven Stories Movie 2 - Side:Blue - Tenrou ...   
943       36699                          Kimitachi wa Dou Ikiru ka   
988       37449                               Strike the Blood III   
1042      36858                                Garo: Uzusumizakura   
1181      20471                                           Aoki Uru   
1298      36587                   Granblue Fantasy The Animation 2   
1310      33852                                  Mekakucity Reload   
1396      35252                                    Hatena☆Illusion   
1409      37396                                        Shikioriori   
1410      37597            Dakaretai Otoko 1-i ni Odosarete Imasu.   
1581      37438                                            Ingress   
1635      37555                         Hangyakusei Million Arthur   
1652      36537                                 Starmyu OVA (2018)   
1702      37806                                   Gunjou no Magmel   
1733      35683                               Midnight Crazy Trail   
1759      37481                                 Ongaku Shoujo (TV)   
1762      32212                      Ensemble Stars! The Animation   
...         ...                                                ...   
14275     37890                Oshi ga Budoukan Ittekuretara Shinu   
14283     37839   Boku wa Robot-goshi no Kimi ni Koi wo Suru Movie   
14284     37760                                          Mecha-ude   
14287     37647                Chibi Airin no Yuruyaka na Nichijou   
14289     37472                          Lalala Lala-chan Season 4   
14290     37340                                 Venus Blood: Brave   
14291     37328                 Oshiete Fairilu: Mahou no Pendulum   
14296     37117                               Bakutsuri Bar Hunter   
14307     37365                           Oshie to Tabi Suru Otoko   
14345     37816                                Bandoe Sareoriratda   
14358     37884                                   Golden Kamuy OVA   
14368     37893                                      Yami Shibai 6   
14371     37904                                         Meshinuma.   
14376     37501           Hug tto! Precure♡Futari wa Precure Movie   
14384     37494                             Chikyuu to no Yakusoku   
14387     37630                                 Washimo 6th Season   
14390     37364                         Kekkai Sensen & Beyond OVA   
14392     37905                                        To the Moon   
14396     37878                                      Sanpou Shoujo   
14404     37827                                        Neko Kikaku   
14408     20263                   W・TAILS CAT - A STRANGE PRESENCE   
14410     37902                     One Piece: Episode of Sorajima   
14414     37662                             Shinya! Tensai Bakabon   
14416     37914                          Chikyuugai Shounen Shoujo   
14417     37915                           Uchuu no Hou: Reimei-hen   
14418     35155  

In [115]:
# determining if there exists unaired values
anime_info_raw['aired'].value_counts() # contains null values
# {'from': None, 'to': None}  
anime_info_raw.loc[(anime_info_raw['aired']== "{'from': None, 'to': None}") & (anime_info_raw['status'] == "Not yet aired"),['anime_id','title','status','aired']].info() 
anime_info_raw.loc[(anime_info_raw['status'] == "Not yet aired"),['anime_id','title','status','aired']] # anime airing date has been announced

# using the status column is a better way to filter unaired anime = those that are still in the works, not yet premiered with only info like production videos, teasers, and the OP having been released  

<class 'pandas.core.frame.DataFrame'>
Int64Index: 147 entries, 76 to 14475
Data columns (total 4 columns):
anime_id    147 non-null int64
title       147 non-null object
status      147 non-null object
aired       147 non-null object
dtypes: int64(1), object(3)
memory usage: 5.7+ KB


anime_id                                              title  \
55        33185           Peace Maker Kurogane Movie 1: Omou Michi   
76        37447                                    Karakuri Circus   
109       36951                            Yuu☆Yuu☆Hakusho Special   
143       37711                                     Koi to Uso OVA   
145       37435                                    Carol & Tuesday   
187       35759                                 Sidonia no Kishi 3   
229       34779                     Ar:pieL - 6-seosgaeui Unmyeong   
334       35503                      Shoujo☆Kageki Revue Starlight   
387       37208                                      Mo Dao Zu Shi   
458       37509                                           W&#039;z   
511       37651     Fate/Extra: Last Encore - Irusterias Tendouron   
639       33248             K: Seven Stories Movie 1 - R:B - Blaze   
671       37106                                        Kori no Koe   
784       37831  Sword Art Online Alternative: Gun Gale Online ...   
864       37779                              Yakusoku no Neverland   
894       35424  Donten ni Warau Gaiden: Shukumei, Soutou no Fuuma   
930       37303  K: Seven Stories Movie 2 - Side:Blue - Tenrou ...   
943       36699                          Kimitachi wa Dou Ikiru ka   
947       36474                      Sword Art Online: Alicization   
988       37449                               Strike the Blood III   
1042      36858                                Garo: Uzusumizakura   
1181      20471                                           Aoki Uru   
1282      34440        Code Geass: Hangyaku no Lelouch III - Oudou   
1298      36587                   Granblue Fantasy The Animation 2   
1310      33852                                  Mekakucity Reload   
1396      35252                                    Hatena☆Illusion   
1409      37396                                        Shikioriori   
1410      37597            Dakaretai Otoko 1-i ni Odosarete Imasu.   
1581      37438                                            Ingress   
1635      37555                         Hangyakusei Million Arthur   
...         ...                                                ...   
14074     33438                            Force of Will the Movie   
14076     37599                                   Magatsu Wahrheit   
14106     37870                             BanG Dream! 3rd Season   
14108     35213                               Kai Feng Kidan Movie   
14184     37873                              BanG Dream! Garu☆PICO   
14227     37286                                  It&#039;s My Life   
14272     37885                                 Dragon Ball Heroes   
14273     37916              Uma Musume: Pretty Derby (TV) Special   
14275     37890                Oshi ga Budoukan Ittekuretara Shinu   
14283     37839   Boku wa Robot-goshi no Kimi ni Koi wo Suru Movie   
14284     37760                                          Mecha-ude   
14289     37472                          Lalala Lala-chan Season 4   
14290     37340                                 Venus Blood: Brave   
14291     37328                 Oshiete Fairilu: Mahou no Pendulum   
14296     37117                               Bakutsuri Bar Hunter   
14307     37365                           Oshie to Tabi Suru Otoko   
14358     37884                                   Golden Kamuy OVA   
14368     37893                                      Yami Shibai 6   
14371     37904                                         Meshinuma.   
14374     37887                                      Huan Jie Wang   
14376     37501           Hug tto! Precure♡Futari wa Precure Movie   
14390     37364                         Kekkai Sensen & Beyond OVA   
14392     37905                                        To the Moon   
14404     37827                                        Neko Kikaku   
14410     37902                     One Piece: Episode of Sorajima   
14414     37662  

#### Analyis: 
No. of rows with status = 'not aired yet' and no aired info ({'from': None, 'to': None}) = 147
whereas the no. of rows with status = 'not aired yet' is much higher at 336 rows -> on inspection, these manga have been completed and their airing info has been announced by the distributor. since we do not have info like score for these manga -> using the condition status = 'not aired yet' is a better filter

In [168]:
anime_info_raw['genre'].isna().value_counts() # cannot predict using genre = null -> drop these anime?
anime_drop_no_genre = anime_info_raw.dropna(subset=['genre'])

In [156]:
anime_info_raw.drop(anime_info_raw.loc[anime_info_raw['status'] == "Not yet aired",:].index).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14142 entries, 0 to 14477
Data columns (total 31 columns):
anime_id          14142 non-null int64
title             14142 non-null object
title_english     5688 non-null object
title_japanese    14111 non-null object
title_synonyms    8780 non-null object
image_url         14050 non-null object
type              14142 non-null object
source            14142 non-null object
episodes          14142 non-null int64
status            14142 non-null object
airing            14142 non-null bool
aired_string      14142 non-null object
aired             14142 non-null object
duration          14142 non-null object
rating            14142 non-null object
score             14142 non-null float64
scored_by         14142 non-null int64
rank              12898 non-null float64
popularity        14142 non-null int64
members           14142 non-null int64
favorites         14142 non-null int64
background        1050 non-null object
premiered         40

In [85]:
anime_extracted = anime_info_raw.drop(columns=['aired_string','title_english','title_japanese', \
                                               'title_synonyms','image_url','type','duration', \
                                               'background','premiered','broadcast','opening_theme',\
                                               'ending_theme','licensor'])
anime_extracted.columns

Index(['anime_id', 'title', 'source', 'episodes', 'status', 'airing', 'aired',
       'rating', 'score', 'scored_by', 'rank', 'popularity', 'members',
       'favorites', 'related', 'producer', 'studio', 'genre'],
      dtype='object')

In [86]:
anime_info_raw['rank'].isna().value_counts() # contains null values

False    12904
True      1574
Name: rank, dtype: int64

In [87]:
anime_info_raw['score'].isna().value_counts() # no null values for score

False    14478
Name: score, dtype: int64

In [88]:
drop_na_rank_score = anime_info_raw.dropna(subset=['rank','score'])
drop_na_rank_score.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12904 entries, 0 to 14477
Data columns (total 31 columns):
anime_id          12904 non-null int64
title             12904 non-null object
title_english     5282 non-null object
title_japanese    12873 non-null object
title_synonyms    8257 non-null object
image_url         12812 non-null object
type              12904 non-null object
source            12904 non-null object
episodes          12904 non-null int64
status            12904 non-null object
airing            12904 non-null bool
aired_string      12904 non-null object
aired             12904 non-null object
duration          12904 non-null object
rating            12904 non-null object
score             12904 non-null float64
scored_by         12904 non-null int64
rank              12904 non-null float64
popularity        12904 non-null int64
members           12904 non-null int64
favorites         12904 non-null int64
background        1036 non-null object
premiered         40

### User list
information about users who watch anime
****


In [89]:
user_anime_list_raw = pd.read_csv('myanimelist/UserAnimeList.csv')
user_anime_list_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80076112 entries, 0 to 80076111
Data columns (total 11 columns):
username               object
anime_id               int64
my_watched_episodes    int64
my_start_date          object
my_finish_date         object
my_score               int64
my_status              int64
my_rewatching          float64
my_rewatching_ep       int64
my_last_updated        int64
my_tags                object
dtypes: float64(1), int64(6), object(4)
memory usage: 6.6+ GB


### User anime list
contains anime lists of all users
****

In [90]:
user_list_raw = pd.read_csv('myanimelist/UserList.csv')
user_list_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 302675 entries, 0 to 302674
Data columns (total 17 columns):
username                    302674 non-null object
user_id                     302675 non-null int64
user_watching               302675 non-null int64
user_completed              302675 non-null int64
user_onhold                 302675 non-null int64
user_dropped                302675 non-null int64
user_plantowatch            302675 non-null int64
user_days_spent_watching    302675 non-null float64
gender                      217800 non-null object
location                    156774 non-null object
birth_date                  168749 non-null object
access_rank                 0 non-null float64
join_date                   302546 non-null object
last_online                 302546 non-null object
stats_mean_score            302546 non-null float64
stats_rewatched             302546 non-null float64
stats_episodes              302546 non-null float64
dtypes: float64(5), int64(6)